File: tests/compiled_workloads/adapters/peer/candidates/embedded-sensor-ring.mjs
import fs from "node:fs";

const raw = fs.readFileSync(process.argv[2], "utf8");
const capacity = 4;
const modulus = 1_000_003;
const queue = [];
let head = 0;
let tail = 0;
let acceptedEnqueues = 0;
let acceptedDequeues = 0;
let rejects = 0;
let digest = 17;
const output = [`capacity=${capacity}`];

function foldDigest(step, code, value) {
  const valuePart = ((value % modulus) + modulus) % modulus;
  digest = (digest * 131 + step * 17 + code * 257 + valuePart * 3 +
    head * 5 + tail * 7 + queue.length * 11) % modulus;
}

function state(step, operation, result, value) {
  const valuePart = value === undefined ? "" : `|value=${value}`;
  output.push(`step=${step}|op=${operation}${valuePart}|result=${result}|head=${head}|tail=${tail}|count=${queue.length}`);
}

let step = 0;
for (const line of raw.split(/\r?\n/)) {
  if (line.length === 0 || line.startsWith("#")) continue;
  step += 1;
  const fields = line.split("|");
  const operation = fields[0];
  const valueText = fields[1];
  const hasExtra = fields.length > 2;
  if (!hasExtra && operation === "ENQ") {
    const value = valueText === undefined || !/^-?\d+$/.test(valueText) ? undefined : Number(valueText);
    if (value === undefined || !Number.isSafeInteger(value)) {
      rejects += 1;
      foldDigest(step, 5, 0);
      state(step, "malformed", "reject-malformed");
    } else if (queue.length >= capacity) {
      rejects += 1;
      foldDigest(step, 3, value);
      state(step, "enqueue", "reject-overflow", value);
    } else {
      queue.push(value);
      tail = (tail + 1) % capacity;
      acceptedEnqueues += 1;
      foldDigest(step, 1, value);
      state(step, "enqueue", "accepted", value);
    }
  } else if (!hasExtra && operation === "DEQ" && valueText === undefined) {
    if (queue.length === 0) {
      rejects += 1;
      foldDigest(step, 4, 0);
      state(step, "dequeue", "reject-underflow");
    } else {
      const value = queue.shift();
      head = (head + 1) % capacity;
      acceptedDequeues += 1;
      foldDigest(step, 2, value);
      state(step, "dequeue", "accepted", value);
    }
  } else {
    rejects += 1;
    foldDigest(step, 5, 0);
    state(step, "malformed", "reject-malformed");
  }
}

output.push(`summary|accepted_enqueues=${acceptedEnqueues}|accepted_dequeues=${acceptedDequeues}|rejects=${rejects}|head=${head}|tail=${tail}|count=${queue.length}`);
output.push(`digest=${digest}`);
process.stdout.write(output.join("\n") + "\n");

File: tests/compiled_workloads/adapters/peer/candidates/embedded-sensor-ring.cpp
#include <array>
#include <charconv>
#include <cstdint>
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <string_view>
#include <vector>

constexpr std::size_t capacity = 4;
constexpr std::int64_t modulus = 1'000'003;

bool parseInteger(std::string_view text, std::int64_t &value) {
    if (text.empty()) return false;
    const auto result = std::from_chars(text.data(), text.data() + text.size(), value);
    return result.ec == std::errc{} && result.ptr == text.data() + text.size();
}

std::int64_t positiveMod(std::int64_t value) {
    value %= modulus;
    return value < 0 ? value + modulus : value;
}

struct Ring {
    std::array<std::int64_t, capacity> slots{};
    std::size_t head = 0;
    std::size_t tail = 0;
    std::size_t count = 0;
    bool enqueue(std::int64_t value) {
        if (count == capacity) return false;
        slots[tail] = value;
        tail = (tail + 1) % capacity;
        ++count;
        return true;
    }
    bool dequeue(std::int64_t &value) {
        if (count == 0) return false;
        value = slots[head];
        head = (head + 1) % capacity;
        --count;
        return true;
    }
};

std::int64_t fold(
    std::int64_t digest,
    std::int64_t step,
    std::int64_t code,
    std::int64_t value,
    const Ring &ring
) {
    return positiveMod(
        digest * 131 + step * 17 + code * 257 + positiveMod(value) * 3 +
        static_cast<std::int64_t>(ring.head) * 5 +
        static_cast<std::int64_t>(ring.tail) * 7 +
        static_cast<std::int64_t>(ring.count) * 11
    );
}

void printState(
    std::vector<std::string> &output,
    std::int64_t step,
    std::string_view operation,
    std::string_view result,
    const Ring &ring,
    const std::int64_t *value = nullptr
) {
    std::ostringstream row;
    row << "step=" << step << "|op=" << operation;
    if (value != nullptr) row << "|value=" << *value;
    row << "|result=" << result << "|head=" << ring.head << "|tail=" << ring.tail
        << "|count=" << ring.count;
    output.push_back(row.str());
}

int main(int argc, char **argv) {
    if (argc < 2) return 2;
    std::ifstream input(argv[1]);
    if (!input) return 2;
    Ring ring;
    std::int64_t step = 0, acceptedEnqueues = 0, acceptedDequeues = 0, rejects = 0, digest = 17;
    std::vector<std::string> output{"capacity=4"};
    std::string line;
    while (std::getline(input, line)) {
        if (line.empty() || line.front() == '#') continue;
        ++step;
        std::vector<std::string_view> fields;
        std::size_t start = 0;
        while (true) {
            const auto separator = line.find('|', start);
            fields.emplace_back(line.data() + start, separator == std::string::npos ? line.size() - start : separator - start);
            if (separator == std::string::npos) break;
            start = separator + 1;
        }
        if (fields.size() == 2 && fields[0] == "ENQ") {
            std::int64_t value = 0;
            if (!parseInteger(fields[1], value)) {
                ++rejects;
                digest = fold(digest, step, 5, 0, ring);
                printState(output, step, "malformed", "reject-malformed", ring);
            } else if (ring.enqueue(value)) {
                ++acceptedEnqueues;
                digest = fold(digest, step, 1, value, ring);
                printState(output, step, "enqueue", "accepted", ring, &value);
            } else {
                ++rejects;
                digest = fold(digest, step, 3, value, ring);
                printState(output, step, "enqueue", "reject-overflow", ring, &value);
            }
        } else if (fields.size() == 1 && fields[0] == "DEQ") {
            std::int64_t value = 0;
            if (ring.dequeue(value)) {
                ++acceptedDequeues;
                digest = fold(digest, step, 2, value, ring);
                printState(output, step, "dequeue", "accepted", ring, &value);
            } else {
                ++rejects;
                digest = fold(digest, step, 4, 0, ring);
                printState(output, step, "dequeue", "reject-underflow", ring);
            }
        } else {
            ++rejects;
            digest = fold(digest, step, 5, 0, ring);
            printState(output, step, "malformed", "reject-malformed", ring);
        }
    }
    std::ostringstream summary;
    summary << "summary|accepted_enqueues=" << acceptedEnqueues
        << "|accepted_dequeues=" << acceptedDequeues << "|rejects=" << rejects
        << "|head=" << ring.head << "|tail=" << ring.tail << "|count=" << ring.count;
    output.push_back(summary.str());
    output.push_back("digest=" + std::to_string(digest));
    for (const auto &row : output) std::cout << row << '\n';
}
